src/main/java/net/minecraftforge/event/entity/living/PotionEvent.java

```diff
@@ -0,0 +1,117 @@
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+package net.minecraftforge.event.entity.living;
+
+import javax.annotation.Nullable;
+
+import net.minecraft.entity.EntityLivingBase;
+import net.minecraft.potion.Potion;
+import net.minecraft.potion.PotionEffect;
+
+import net.minecraftforge.fml.common.eventhandler.Cancelable;
+import net.minecraftforge.fml.common.eventhandler.Event.HasResult;
+
+/**
+ * This Event and its subevents gets fired from  {@link EntityLivingBase} on the  {@link MinecraftForge#EVENT_BUS}.<br>
+ */
+public class PotionEvent extends LivingEvent
+{
+    @Nullable
+    public final PotionEffect effect;
+    
+    public PotionEvent(EntityLivingBase living, PotionEffect eff)
+    {
+        super(living);
+        effect = eff;
+    }
+    
+    public PotionEffect getPotionEffect()
+    {
+        return effect;
+    }
+    
+    /**
+     * This Event is fired when a Potion is about to get removed from an Entity
+     * This Event is {@link Cancelable}
+     * This Event does not have a result.
+     */
+    @Cancelable
+    public static class PotionRemoveEvent extends PotionEvent
+    {
+        private final Potion toRemove;
+        
+        public PotionRemoveEvent(EntityLivingBase living, Potion  potion)
+        {
+            super(living, living.getActivePotionMap().get(potion));
+            toRemove = potion;
+        }
+        
+        public Potion getPotion()
+        {
+            return toRemove;
+        }
+    }
+    
+    /**
+     * This Event is fired to check if a Potion can get applied to an Entity.
+     * This Event is not {@link Cancelable}
+     * This Event has a result {@link HasResult}.
+     * ALLOW will apply this potion effect
+     * DENY will not apply this potion effect
+     * DEFAULT will run vanilla logic to determine if this potion isApplicable
+     */
+    @HasResult
+    public static class PotionApplicableEvent extends PotionEvent
+    {
+        public PotionApplicableEvent(EntityLivingBase living, PotionEffect eff)
+        {
+            super(living, eff);
+        }
+    }
+    
+    /**
+     * This Event is fired when a new Potion is added to the Entity.
+     * This Event is not {@link Cancelable}
+     * This Event does not have a Result.
+     */
+    public static class NewPotionEvent extends PotionEvent
+    {
+        public NewPotionEvent(EntityLivingBase living, PotionEffect eff)
+        {
+            super(living, eff);
+        }
+    }
+    
+    /**
+     * This Event is fired when a Potion is changed.
```

Comment:
what does "changed" mean?

Reply:
chnaged is when you have a new potion effect of the same kind but with different level or time, so they get merged.

Reply:
I'm on the border of whether this one is useful or if it's just polluting the interface. When would you want to distinguish this and a normal add?

Anyhow, should be named consistently with remove: `PotionChangedEvent`

Reply:
Over at #4375 it seems like they whould disrtinguish them, and it make sense if you only want to do something if its realy a new effect.

Reply:
Why not merge this into the potion added event, and just add a nullable existingEffect/ oldEffect?

Reply:
Well in order to get the old effect I would need to do alot of patches in the EntityLivingBase